### PR TITLE
Modified installation script to pull .exe on Windows / Git Bash

### PR DESCRIPTION
**Pull request by wsteimel77**
## Description
Added another case when parsing uname to try and get the correct binary suffix when installing on a Windows system and retain the .exe extension when saving to temp dir.

## Motivation and Context
Running the script in Windows currently installs the Linux binary
- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))
#159 

## How Has This Been Tested?
Invoked in Git Bash on Windows 10:
```
$ sh get.sh
You already have the faas-cli!
Overwriting in 1 seconds.. Press Control+C to cancel.
Getting package https://github.com/openfaas/faas-cli/releases/download/0.4.18b/faas-cli.exe
Attemping to move faas-cli to /usr/local/bin
New version of faas-cli installed to /usr/local/bin
$ faas-cli version
  ___                   _____           ____
/ _ \ _ __   ___ _ __ |  ___|_ _  __ _/ ___|
| | | | '_ \ / _ \ '_ \| |_ / _` |/ _` \___ \
| |_| | |_) |  __/ | | |  _| (_| | (_| |___) |
\___/| .__/ \___|_| |_|_|  \__,_|\__,_|____/
      |_|

Commit: e804a60889e4d339762fb3643a315a173a963f26
Version: 0.4.18b
```

Then tested in Powershell with faas-cli.exe included in the path and was able to successfully invoke faas-cli because .exe is retained.

I have also successfully updated faas-cli on Arch Linux using the revised script. 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: get.sh [hunk 1]
```diff
@@ -21,6 +21,7 @@ getPackage() {
     uname=$(uname)
 
     suffix=""
+    temp="/tmp/faas-cli"
```
Comment (alexellis) [inline get.sh]: Can we call this something else such as CLI ($CLI)?

**Comment by alexellis** [comment]
@wsteimel77 thanks for the PR. Think this will really help Windows users.

**Comment by johnmccabe** [comment]
Hi @wsteimel77 how does this behave when called from WSL?

Also is the `.exe` needed when being used with Git bash?

File: get.sh [hunk 2]
```diff
@@ -39,21 +40,25 @@ getPackage() {
...
 
-    curl -sSL $url > /tmp/faas-cli
+    curl -sSL $url > $temp
 
     if [ "$?" = "0" ]; then
         echo "Attemping to move faas-cli to /usr/local/bin"
-        chmod +x /tmp/faas-cli
-        cp /tmp/faas-cli /usr/local/bin/
+        chmod +x $temp
+        cp $temp /usr/local/bin/
```
Comment (johnmccabe) [inline get.sh]: `/usr/local/bin/` is a non-standard location on windows - its not part of the Mingw64 Git Bash install - how do you see this being created on windows?

Comment (johnmccabe) [reply get.sh]: For example, running on Windows 10:
```
$ ./get.sh
which: no faas-cli in (/c/Users/John McCabe/bin:/mingw64/bin:/usr/local/bin:/usr/bin:/bin:/mingw64/bin:/usr/bin:/c/Users/John McCabe/bin:/c/Program Files/Docker/Docker/Resources/bin:/d/Oculus/Support/oculus-runtime:... snip ...)
Getting package https://github.com/openfaas/faas-cli/releases/download/0.4.19/faas-cli.exe
Attemping to move faas-cli to /usr/local/bin
cp: cannot create regular file '/usr/local/bin/': No such file or directory
```

Comment (wsteimel77) [reply get.sh]: @johnmccabe, I hadn't tried it on WSL.  I will give that a go and let you know.  The .exe extension is not needed when just using from Git bash, but is a problem if you try to use it from cmd or Powershell, for instance.  And yes, /usr/local/bin is not a standard location, I just wasn't sure how much to change at once or where it should be placed.

Comment (wsteimel77) [reply get.sh]: Within WSL, uname returns Linux, so it actually installs the Linux binary, and that works correctly within WSL.

Comment (wsteimel77) [reply get.sh]: @johnmccabe, I just tried an install of Git Bash on a new Windows machine, and it appears that /usr/local/bin is part of the default path already, it just hasn't been created yet.  We could do mkdir -p on the script to create if it doesn't exist?  What are your thoughts?

Comment (alexellis) [reply get.sh]: Where are bin files stored normally with Git Bash?

**Comment by wsteimel77** [comment]
It appears that some things are in /bin, some are in /mingw64/bin and some
are in /usr/bin.  /usr/local/bin is part of the default search path in Git
Bash, it just isn't created during install.  @johnmccabe, do you concur?

On 26 Oct 2017 07:52, "Alex Ellis" <notifications@github.com> wrote:

*@alexellis* commented on this pull request.
------------------------------

In get.sh
<https://github.com/openfaas/faas-cli/pull/171#discussion_r147057594>:

>
     if [ "$?" = "0" ]; then
         echo "Attemping to move faas-cli to /usr/local/bin"
-        chmod +x /tmp/faas-cli
-        cp /tmp/faas-cli /usr/local/bin/
+        chmod +x $temp
+        cp $temp /usr/local/bin/

Where are bin files stored normally with Git Bash?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openfaas/faas-cli/pull/171#discussion_r147057594>, or mute
the thread

File: get.sh (hunk 2 above)
Comment (wsteimel77) [reply get.sh]: It appears that some things are in /bin, some are in /mingw64/bin and some
are in /usr/bin.  /usr/local/bin is part of the default search path in Git
Bash, it just isn't created during install.  @johnmccabe, do you concur?

Comment (wsteimel77) [reply get.sh]: A few examples of what I see
```
WESTON@FAAS-WIN MINGW64 ~
$ which curl
/mingw64/bin/curl

WESTON@FAAS-WIN MINGW64 ~
$ which git
/mingw64/bin/git

WESTON@FAAS-WIN MINGW64 ~
$ which ssh
/usr/bin/ssh

WESTON@FAAS-WIN MINGW64 ~
$ which awk
/usr/bin/awk
```

**Comment by alexellis** [comment]
Is /usr/local/bin the correct location for Git Bash? I wonder what @shanselman thinks.. in terms of installing with a shell-script on Windows?

* On WSL you can use the Linux binary
* When running the script through mini-gw/git bash we have to determine to download the faas-cli.exe

We also appear to have a chocolatey package underway in the linked issue 👍

**Comment by shanselman** [comment]
There is no concept of /usr/local/bin on Windows, nor is there a standard "all this user stuff is in the PATH." Sadly Chocolatey is the only real way to solve this today. Folks will not understand you making a /user/local/bin for them.

**Comment by wsteimel77** [comment]
So does it seem best to just point windows users to the windows binary from the documentation and tell them to add it to their path if not using Chocolatey, or would it be worth it at all to try building a Windows installer?  It looks like minikube, for instance, has an experimental windows installer that could possibly serve as a starting point.

**Comment by johnmccabe** [comment]
@wsteimel77 / @alexellis should we close this out and set #159 to `revisit`?

I'm not a fan of the installer personally, this is very much a dev tool and I think we're probably safe assuming a dev is comfortable with adding stuff to the path?

Personally I have a single dir `D:\localbin` where I drop everything I need to add to the path.

**Comment by wsteimel77** [comment]
@johnmccabe, yeah, I thought the same as well, but just wanted to make sure it wasn't just me.

**Comment by alexellis** [comment]
Ok maybe we can just output a comment for windows with the release URL?

**Pull request closed without merging**